src/backend/access/appendonly/appendonlyam_handler.c

```diff
@@ -545,8 +545,12 @@ appendonly_index_fetch_tuple(struct IndexFetchTableData *scan,
 	 */
 	Assert(aoscan->aofetch->snapshot == snapshot);
 
+	aoscan->aofetch->visibilityMap.visimapStore.snapshot = snapshot;
+
 	appendonly_fetch(aoscan->aofetch, (AOTupleId *) tid, slot);
 
+	aoscan->aofetch->visibilityMap.visimapStore.snapshot = InvalidSnapshot;
+
 	return !TupIsNull(slot);
```

Comment:
Hmm.. set/reset at fetching each index tuple, we usually have million/billion of rows, and the set/rest cost will be double.
That's not great.

Reply:
Thanks for your suggestion, I think it is not great, too. But it works....
1. we have already try another easy way to solve this problem, by removing register snapshot. But failed in CI test and trying to fix it. https://github.com/cloudberrydb/cloudberrydb/pull/136
2. we see the same way to slove the problem in unique index check by set/reset. And I think AO may not efficent in index fetch, so millions tuple index scan may not efficient for index fetch even without set and reset 
3. beg your better way to solve this problem, thanks.